community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/SpatialFusionSchemaIndexProvider.java

```diff
@@ -0,0 +1,239 @@
+            }
+        }
+        return state;
+    }
+
+    @Override
+    public IndexCapability getCapability( IndexDescriptor indexDescriptor )
+    {
+        // Spatial indexes are not ordered, nor do they return complete values
+        return IndexCapability.NO_CAPABILITY;
+    }
+
+    @Override
+    public StoreMigrationParticipant storeMigrationParticipant( FileSystemAbstraction fs, PageCache pageCache )
+    {
+        // Since this native provider is a new one, there's no need for migration on this level.
+        // Migration should happen in the combined layer for the time being.
+        return StoreMigrationParticipant.NOT_PARTICIPATING;
+    }
+
+    @Override
+    public SpatialKnownIndex selectAndCreate( Map<CoordinateReferenceSystem,SpatialKnownIndex> indexMap, long indexId, Value value )
+    {
+        PointValue pointValue = (PointValue) value;
+        CoordinateReferenceSystem crs = pointValue.getCoordinateReferenceSystem();
+        return selectAndCreate( indexMap, indexId, crs );
+    }
+
+    @Override
+    public SpatialKnownIndex selectAndCreate( Map<CoordinateReferenceSystem,SpatialKnownIndex> indexMap, long indexId, CoordinateReferenceSystem crs )
+    {
+        return indexMap.computeIfAbsent( crs,
+                k -> new SpatialKnownIndex( directoryStructure(), crs, indexId, pageCache, fs, monitor, recoveryCleanupWorkCollector ) );
+    }
+
+    private void findAndCreateKnownSpatialIndexes()
+    {
+        Pattern pattern = Pattern.compile( "(\\d+)-(\\d+)" );
+        File rootDirectory = this.directoryStructure().rootDirectory();
+        if ( rootDirectory == null )
+        {
+            return;
+        }
+        File[] files = rootDirectory.listFiles();
+        if ( files != null )
+        {
+            for ( File file : files )
+            {
+                if ( file.isDirectory() )
+                {
+                    Integer indexId = Integer.parseInt( file.getName() );
+                    File[] subdirs = this.directoryStructure().directoryForIndex( indexId ).listFiles();
+                    if ( subdirs != null )
+                    {
+                        for ( File subdir : subdirs )
+                        {
+                            Matcher m = pattern.matcher( subdir.getName() );
+                            if ( m.matches() )
+                            {
+                                int tableId = Integer.parseInt( m.group( 1 ) );
+                                int code = Integer.parseInt( m.group( 2 ) );
+                                CoordinateReferenceSystem crs = CoordinateReferenceSystem.get( tableId, code );
+                                SpatialKnownIndex index = selectAndCreate( indexesFor( indexId ), indexId, crs );
+                                if ( !index.indexExists() )
+                                {
+                                    // TODO: log something here?
+                                    indexesFor( indexId ).remove( crs );
```

Comment:
We were thinking there are only two states, the index exists, or it doesn't. Are you suggesting we try to deal with cases where, for some reason, a directory has been made (or an index deleted, but leaving directories)?

I had assumed a FAILED state was associated with an index that actually did exist, but the contents were broken for some reason (out of disk-space or power failure during populating perhaps). The case where some directories exist, bt not the index itself, appears to me to be a non-existing index.

Reply:
This is how I think it works, please correct me if I'm wrong here 🗡 
1. We only create a directory for a given CRS when we actually insert some value there.
2. We never delete index files, even if they are empty.
Conclusion, if there is a directory there must be an index file. And if there is not, there is no way for us to determine if that file had some data that is now missing or not.
We expect data to be their, it is not, thus I think this would be considered as FAILED.